File: src/clojure/writer.clj
(ns writer
  (:use util las))

(import '(java.io File FileWriter BufferedWriter))

(load "parser")
(refer 'private-parser)
(refer 'parser)

(def writer)

(defn write-data [lasfile]
  (.write writer "~Ascii \n")
  (let [curves (conj (:curves lasfile) (:index lasfile))
	columns (count curves)
	rows (count (:data (first curves)))
	row-data (fn [r] (map #(format "%1.20f " (bigdec (nth (:data %) r))) curves))]
    (doseq [r (range 0 rows)]
      (doto writer
	(.write (seq-to-str (row-data r)))
	(.newLine)))))		

(defn write-descriptor [descriptor]
  (doto writer
    (.write (str (:mnemonic descriptor) " ."))
    (.write (str (:unit descriptor) " "))
    (.write (str (:data descriptor) " : "))
    (.write (str (:description descriptor)))
    (.newLine)))

(defn write-header [header]
  (let [type (:type header)
	descriptors (:descriptors header)
	prefix (get header-prefixes type)]
    (doto writer
      (.write (str prefix))
      (.newLine))
    (doseq [descriptor descriptors]
      (write-descriptor descriptor))))

(defn write-headers [lasfile]
  (let [headers [:version-header
		 :well-header
		 :curve-header
		 :parameter-header]]
    (doseq [h headers]
      (write-header (get lasfile h)))))

(defmulti write-las-file (fn [x y] (class x)))

(defmethod write-las-file File [file-handle lasfile] 
  (let [file-writer (new BufferedWriter (new FileWriter file-handle))]
    (binding [writer file-writer]
      (write-headers lasfile)
      (write-data lasfile))
    (. file-writer (close))))

(defmethod write-las-file String [path lasfile]
  (let [file-handle (new File path)]
    (write-las-file file-handle lasfile)))

File: src/scala/core/writer.scala
package core

import scala.collection.jcl.Conversions._
import java.io._
import java.util.{List,Formatter,LinkedList}
import java.math.BigDecimal

object DefaultLasWriter extends LasWriter {

  def writeLasFile(lf: LasFile, file:File) { 
    val writer = new BufferedWriter(new FileWriter(file))
    try {
      writeHeaders(lf, writer)
      writeCurves(lf, writer)
    } finally {
      writer.close()
    }
  }

  def writeLasFile(lf: LasFile, path:String) { 
    writeLasFile(lf, new File(path))
  }

  private def writeHeaders(lf: LasFile, writer:BufferedWriter) {
    val headers = lf.getHeaders
    for(h <- headers) {
      writeHeader(h, writer)
    }
  }

  private def writeHeader(h:Header, writer:BufferedWriter) {
    writer.write(h.getPrefix); writer.newLine
    writeDescriptors(h.getDescriptors,
		     writer)
  }

  private def writeDescriptors(descriptors:List[Descriptor], writer: BufferedWriter){
    for(d <- descriptors){
      writeDescriptor(d, writer)
    }
  }

  private def writeDescriptor(descriptor:Descriptor, writer: BufferedWriter) {
    val write = (s:String) => writer.write(s)
    write(descriptor.getMnemonic)
    write(" .")
    write(descriptor.getUnit.toString)
    write(" ")
    write(descriptor.getData.toString)
    write(" : ")
    writer.newLine
  }

  private def writeCurves(lf: LasFile, writer: BufferedWriter) {
    writer.write("~A")
    writer.newLine
    val curves = new LinkedList[Curve](lf.getCurves)
    curves.add(lf.getIndex)
    val columns = curves.size
    val rows = curves.first.getLasData.size
    def row_data(r:Int) = curves.map(c => {
      val data = c.getLasData
      val point = data.get(r).asInstanceOf[BigDecimal]
      val precision = Math.max(point.precision,10)
      String.format("%1." + precision + "f",Array(point))
    })

    for(r <- 0 until rows){
      writer.write(row_data(r).mkString(" "))
      writer.newLine
    }
  }			     
    
}
